doc/debugging-dmtcp.txt

```diff
@@ -2,35 +2,128 @@ This is intended as one of a series of informal documents to describe
+Note that GDB will first give you control under 'dmtcp_launch'.  
+(1) If you want to stop GDB when dmtcp1 enters its main() routine:
+  (gdb) break execvp
+  (gdb) run
+  (gdb) break main
+  (gdb) continue
+(2) If you want to stop GDB when the DMTCP constructor takes control
+      in dmtcp1 (even before entering the main() routine of dmtcp1), do:
+  (gdb) break 'dmtcp::DmtcpWorker::DmtcpWorker()'
+    [ and say yes to:  "breakpoint pending on future shared library load" ]
+  (gdb) run
 
 ===
+D. DEBUGGING ON RESTART USING GDB ATTACH
+
+Ideally, we would like to just use:  gdb --args dmtcp_restart ckpt_*.dmtcp
+This does not work (unless you just want to debug the command dmtcp_restart
+itself).  But the command dmtcp_restart overlays the old memory from
+ckpt_*.dmtcp into the current running 'dmtcp_restart' process.  As you would
+expect, GDB has difficulty following this memory overlay.  Therefore,
+the best way to debug a restarted process is to use the 'attach' feature
+of GDB.
+
+If you want to debug DMTCP functions also, be sure to compile DMTCP with
+debugging support.  A convenient recipe from the root directory of DMTcP is:
+  ./configure CXXFLAGS="-g3 -O0" CFLAGS="-g3 -O0" 
+  make -j clean
+  make -j
+
+There are four methods available.  Most users will prefer the first,
+or possibly the second method.  The methods are in order of increasing
+interest for developers of DMTCP.  Make sure that you compiled DMTCP
+with debugging support if you want to also see symbols for DMTCP functions.
+
+1. DMTCP_GDB_ATTACH_ON_RESTART:
+2. DMTCP_RESTART_PAUSE2:
+3. DMTCP_RESTART_PAUSE:
+4. USING GDB WITH MTCP (primarily for developers):
+
+The four methods follow:
+
+1. DMTCP_GDB_ATTACH_ON_RESTART:
+In the past, using 'gdb attach' was as easy as:
+       dmtcp_restart ckpt_a.out_*.dmtcp &
+       gdb a.out `pgrep -n a.out`
+
+Unfortunately, recent Linux distros have forbidden this type of
+'gdb attach' because it can expose security holes.  To get around
+this on a per-process basis, DMTCP provides the special
+environment variable DMTCP_GDB_ATTACH_ON_RESTART:
+       DMTCP_GDB_ATTACH_ON_RESTART=1 dmtcp_restart ckpt_a.out_*.dmtcp &
+       gdb a.out `pgrep -n a.out`
+
+2. DMTCP_RESTART_PAUSE2:
+If it's important to attach with GDB early in the restart process, then
+DMTCP provides this and the following alternatives:
+       DMTCP_RESTART_PAUSE2=1 bin/dmtcp_launch a.out
+       bin/dmtcp_command --checkpoint
+       dmtcp_restart ckpt_a.out_*.dmtcp &
+       # DMTCP will then pause for 15 seconds, and wait for a GDB attach.
+
+3. DMTCP_RESTART_PAUSE:
+If you need to attach with GDB _very_ early in the restart process (usually
+needed only by DMTCP developers), do:
+       DMTCP_RESTART_PAUSE=1 bin/dmtcp_launch a.out
+       # Continue as above.
+In this case, you will attach at a time when only the primary thread
+exists, and the DMTCP checkpoint thread is not yet distinguished from
+the original user thread executing main().
```

Comment:
DMTCP_RESTART_PAUSE2 pauses DMTCP immediately after all the threads (including the checkpoint thread) have been restored?